### PR TITLE
enable utils_format.py and load.gtfsfeed_to_df to complete successfully

### PR DESCRIPTION
**Pull request by kuanb**
As I attempted to resolve the issue I noted in [this PR](https://github.com/UDST/urbanaccess/issues/10), I "unravelled" a number of components within `utils_format.py` that were broken. This PR fixes the errors insofar as it assumes that resolving the errors does not negatively affect the quality or accuracy of the output dataframes.

I noticed there is a Travis file. Also, looking through the tests, it looks like this segment my PR covers is missing tests. I'll make an effort to produce some tests to cover this component as well so that, in the future, modifications can be made on these components with confidence.

That said, I wanted to open the PR right now to "get the discussion going" and also ensure that my modifications were not in anyway creating observable negative downstream side effects (i.e. rows being left in the dataframes, etc.).

Issues being resolved in this PR:

1. In `calendar_dates_agencyid`, the merge with `tmp2` requires `service_id`, yet it is not provided in the prior dataframes used to create that variable.

2. Resolving that unveils another issue: The subsitution step in `calendar_dates_agencyid` includes the conversion of spaces to underscores: `sub(r'\s+', '_', col)`. This errors when performed on a dataframe column. I note that there are examples showing this is doable, perhaps it is not supported on some versions of Pandas. Because we don't specify exactly which version of each dependency we use (the greater than or equal option is used in the `requires.txt`), we should opt for a more conservative method in this step.

3. Resolving that step reveals a number of attempted merges on Series instead of dataframes. For example, this line in `calendar_agencyid`: `merged_df = pd.merge(calendar_df['service_id'], tmp2, how='left', on='service_id', sort=False, copy=False)`. `calendar_df['service_id']` returns a Pandas Series. It needs to be nested once more to retain the dataframe.

4. Line 426-5 has the same issue of their not being a `service_id` included in the initial few dataframes, as well.

5. `stop_times_agencyid` function attempts to merge a Series (`trips_df['trip_id']`) with a DataFrame, as well. 

6. Also, issue here (with same code from Item 5) is that you can't left join on `route_id` if it is not included at all in one of the dataframes.

7. The `elif len(agency_df['agency_name']) > 1:` logic segment never udpates the overall `df_list` global, resulting in a `KeyError` when looking for `unique_agency_id` in the next step.

8. General `TODO`s identifying brittle points in the logic.

### CODE REVIEW

File: urbanaccess/gtfs/headways.py [hunk 1]
```diff
@@ -129,10 +129,15 @@ def headways(gtfsfeeds_df,headway_timerange):
         route stop headways in units of minutes
         with relevant route and stop information
     """
+    
+    # TODO: Assertions in code during runtime should be handled in some other way.
```
Comment (sablanchard) [inline urbanaccess/gtfs/headways.py]: Agreed the assertions could be replaced with exceptions. We can address this later.

File: urbanaccess/gtfs/load.py [hunk 2]
```diff
@@ -227,6 +227,9 @@ def gtfsfeed_to_df(gtfsfeed_path=None,validation=False,verbose=True,bbox=None,re
                                                        trips_df=trips_df[['trip_id','route_id']],
                                                        info_to_append='route_type_to_stop_times')
 
+        # TODO: Appending seems improper; given the initial dataframes are empty we should either override or
```
Comment (sablanchard) [inline urbanaccess/gtfs/load.py]: The initial empty dataframes are created in order to use the append functions here. Append is nice in that dfs with different columns do not throw errors. The dfs being appended are all different dfs in a loop from different gtfs sources with structures that may vary. If you have suggestions for another method to go about this please let us know, we would be happy to test it out.

File: urbanaccess/gtfs/utils_format.py [hunk 3]
```diff
@@ -388,12 +389,16 @@ def add_unique_agencyid(agency_df=None,stops_df=None,routes_df=None,trips_df=Non
...
-    if ((os.path.exists(os.path.join(feed_folder,'agency.txt')) == False or
-                 'agency_id' not in agency_df.columns) and
-                nulls_as_folder == True):
-        for df in df_list:
+    path_absent = os.path.exists(os.path.join(feed_folder,'agency.txt')) == False
+    agency_absent = 'agency_id' not in agency_df.columns
+    if ((path_absent or agency_absent) and nulls_as_folder == True):
+
+        for index, df in enumerate(df_list):
+            # TODO: We seem to be repeating this pattern in a number of places - either do it once or use a helper function
```
Comment (sablanchard) [inline urbanaccess/gtfs/utils_format.py]: This TODO refers to a specific unique_agency_id generation that relies on the feed folder name, not a dataframe column name. This is separate from the column agency name function you created below. I am ok with having this outside a helper function but if you wanted to include this in your function below that is ok, the scope would just increase.

File: urbanaccess/gtfs/utils_format.py [hunk 4]
```diff
@@ -406,23 +411,34 @@ def add_unique_agencyid(agency_df=None,stops_df=None,routes_df=None,trips_df=Non
 
         if len(agency_df['agency_name']) == 1:
             assert agency_df['agency_name'].isnull().values == False
+
+            # TODO: Again, this need to be moved into a helper function
```
Comment (sablanchard) [inline urbanaccess/gtfs/utils_format.py]: This TODO refers to a specific unique_agency_id generation that relies on the agency column at index 0 which is different from the agency name function you created below. However, this TODO can stay here because it would be good to add a parameter to your function to accommodate a call to a index so that this can be folded into your function.

File: urbanaccess/gtfs/utils_format.py [hunk 5]
```diff
@@ -406,23 +411,34 @@ def add_unique_agencyid(agency_df=None,stops_df=None,routes_df=None,trips_df=Non
...
             unique_agency_id = sub(r'\s+', '_', agency_df['agency_name'][0]).replace('&','and').lower()
-            for df in df_list:
+
+            for index, df in enumerate(df_list):
                 df['unique_agency_id'] = unique_agency_id
+                df_list[index] = df
             log('The unique agency id: {} was generated using the name of the agency in the agency.txt file.'.format(unique_agency_id))
 
         elif len(agency_df['agency_name']) > 1:
+            # TODO: Assertions shouldn't be in runtime - validation should
```
Comment (sablanchard) [inline urbanaccess/gtfs/utils_format.py]: Yes agreed this assertion can be an exception. I will make the change apart from this PR for this one.

File: urbanaccess/gtfs/utils_format.py [hunk 6]
```diff
@@ -406,23 +411,34 @@ def add_unique_agencyid(agency_df=None,stops_df=None,routes_df=None,trips_df=Non
...
+                df_list[index] = df
             log('The unique agency id: {} was generated using the name of the agency in the agency.txt file.'.format(unique_agency_id))
 
         elif len(agency_df['agency_name']) > 1:
+            # TODO: Assertions shouldn't be in runtime - validation should 
+            #       either be prior to model execution or handled gracefully
+            #       through caught errors/exceptions
             assert agency_df[['agency_id','agency_name']].isnull().values.any() == False
 
+            # TODO: In each of the steps, the functions foo_agencyid ought be prepended with an underscore (e.g.
```
Comment (sablanchard) [inline urbanaccess/gtfs/utils_format.py]: This is a great point. The entire repo should reflect this. We will make a separate PR that adds the underscore to functions that should not be used outside their respective .py files.

File: urbanaccess/gtfs/utils_format.py [hunk 7]
```diff
@@ -441,15 +457,25 @@ def add_unique_agencyid(agency_df=None,stops_df=None,routes_df=None,trips_df=Non
                                                 trips_df=trips_df[['trip_id', 'route_id']],
                                                 agency_df=agency_df[['agency_id','agency_name']])
 
+            # TODO: It's obfuscatory to update the dataframe variables in such a deeply nested way.
```
Comment (sablanchard) [inline urbanaccess/gtfs/utils_format.py]: Yes agreed! Your changes are more clear. Please remove this TODO from this PR since if you dont mind as you have fixed the issue.

File: urbanaccess/gtfs/utils_format.py [hunk 8]
```diff
@@ -441,15 +457,25 @@ def add_unique_agencyid(agency_df=None,stops_df=None,routes_df=None,trips_df=Non
...
+
+            # need to update the df_list object with these new variable overrides
+            df_list = [stops_df,routes_df,trips_df,stop_times_df,calendar_df,calendar_dates_df]
+            
             log('agency.txt agency_name column has more than one agency name listed. Unique agency id was assigned using the agency id and associated agency name.')
 
-    for df in df_list:
+    for index, df in enumerate(df_list):
         if df['unique_agency_id'].isnull().values.any():
+            # TODO: These string conversions seem to follow a pattern, could be part of the helper function?
```
Comment (sablanchard) [inline urbanaccess/gtfs/utils_format.py]: This TODO refers to a specific unique_agency_id generation that relies on the feed folder name, not a dataframe column name. This is separate from the column agency name function you created below. I am ok with having this outside a helper function but if you wanted to include this in your function below that is ok, the scope would just increase.

File: urbanaccess/gtfs/utils_format.py [hunk 9]
```diff
@@ -691,3 +717,12 @@ def append_route_type(stops_df=None, stop_times_df=None,routes_df=None,trips_df=
 
         return stop_times_df
 
+# helper/utility functions
```
Comment (sablanchard) [inline urbanaccess/gtfs/utils_format.py]: This is great!

File: urbanaccess/gtfs/headways.py (hunk 1 above)
Comment (kuanb) [inline urbanaccess/gtfs/headways.py]: For sure, just wanted to place flags as reminders for later :)

File: urbanaccess/gtfs/utils_format.py (hunk 3 above)
Comment (kuanb) [inline urbanaccess/gtfs/utils_format.py]: I think discrete, small helper functions would be fine (making another for this case, that is). Can submit a PR for all these small items later.

Comment (sablanchard) [inline urbanaccess/gtfs/utils_format.py]: 👍

**Comment by kuanb** [comment]
Thanks for the review, https://github.com/UDST/urbanaccess/pull/13/commits/5379f4716e2915da75540563f6bb23e095965a86 introduces the requested modifications (remove TODO, implement suggestion) and also reduces subsetting operations on dataframes to make the code more DRY in that section.

Ready for a squash and merge if nothing else remains!

File: urbanaccess/gtfs/utils_format.py [hunk 10]
```diff
@@ -426,43 +426,58 @@ def add_unique_agencyid(agency_df=None,stops_df=None,routes_df=None,trips_df=Non
...
+
+            stop_times_replacement_df = stop_times_agencyid(
+                                            stop_times_df=stop_times_df,
+                                            routes_df=subset_routes_df,
+                                            trips_df=subset_trips_df,
+                                            agency_df=subset_agency_df)
 
             # need to update the df_list object with these new variable overrides
-            df_list = [stops_df,routes_df,trips_df,stop_times_df,calendar_df,calendar_dates_df]
+            df_list = [stops_replacement_df,
```
Comment (sablanchard) [inline urbanaccess/gtfs/utils_format.py]: This looks good but these new replacement dfs should be returned outside the function - in this PR the old dfs are being returned. Can you make the change to the return statement on this line: https://github.com/kuanb/urbanaccess/blob/5379f4716e2915da75540563f6bb23e095965a86/urbanaccess/gtfs/utils_format.py#L496 ?

Comment (kuanb) [inline urbanaccess/gtfs/utils_format.py]: Ah good point!! Whoops. 

How about an "unpack" loop where we extract each of the dfs into a tuple?

Comment (kuanb) [inline urbanaccess/gtfs/utils_format.py]: Seems like the pythonic way would be to "not care" so might opt for the list since it removes an unnecessary step http://stackoverflow.com/questions/8159107/python-return-tuple-or-list

Comment (sablanchard) [inline urbanaccess/gtfs/utils_format.py]: Thanks! Are there advantages to changing it in that manner?
